### ISSUE TITLE
Microsoft.Extensions.Hosting.IHost should be IAsyncDisposable 

### ISSUE THREAD

**Issue by JanEggers** [dotnet/runtime#43027]
Id like to be able to use the async dispose capabilities of the host without casting 

Proposed API

https://github.com/dotnet/runtime/blob/bd33fc4e4a3382422bee156cd66a45a13521ec58/src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/IHost.cs#L13

=> 
public interface IHost : IDisposable, IAsyncDisposable 

## Usage Examples

``` C#
await using host = new HostBuilder().Build();
```

instead of 

``` C#
var host = new HostBuilder().Build();
await using (host as IAsyncDisposable);
```

**Comment by eerhardt** [comment]
This would be an API breaking change. @davidfowl - do we have a policy on breaking changes like this in Microsoft.Extensions?

**Comment by davidfowl** [comment]
> This would be an API breaking change. @davidfowl - do we have a policy on breaking changes like this in Microsoft.Extensions?

Yep. We'd like to avoid breaking the ecosystem. If I were to guess though, this wouldn't break many people

**Comment by kimsey0** [comment]
Now that we have default interface methods, can't we add this in a non-breaking way?
```csharp
ValueTask IAsyncDisposable.DisposeAsync()
{
    this.Dispose();
    return ValueTask.CompletedTask;
}
```

**Comment by kimsey0** [comment]
Yes, and I propose that that can be achieved in a non-breaking way by having the `IHost` interface implement the `IAsyncDisposable` interface with a [default implementation](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#default-interface-methods) for the `DisposeAsync` method by adding the above-mentioned snippet to `IHost`. 🙂

**Comment by eerhardt** [comment]
@bartonjs can correct me if I'm wrong, but I believe this is the one rule we know about default interface methods - you can't use them in an interface to provide an implementation for another interface's method. When you do this, you can cause diamond issues.

**Comment by kimsey0** [comment]
I didn't know this rule, but it makes sense. Implementing a method from another interface doesn't seem to be disallowed, but if you implement two interfaces which both have default implementations of a third interface's method and don't yourself implement that method, you get a an error:

```csharp 
interface IA : IAsyncDisposable { ValueTask IAsyncDisposable.DisposeAsync() => ValueTask.CompletedTask; }
interface IB : IAsyncDisposable { ValueTask IAsyncDisposable.DisposeAsync() => ValueTask.CompletedTask; }
interface IAB : IA, IB { }

class AB : IAB { }
```
gives `CS8705 Interface member 'IAsyncDisposable.DisposeAsync()' does not have a most specific implementation. Neither 'IA.IAsyncDisposable.DisposeAsync()', nor 'IB.IAsyncDisposable.DisposeAsync()' are most specific.`

In the concrete case of `IHost`, I guess this is breaking if a consumer has a class which implements `IHost` and another interface with a default implementation of `IAsyncDisposable.DisposeAsync`. However, I guess that will still break fewer people than adding `IAsyncDisposable` to `IHost` without the default implementation.

**Comment by bartonjs** [comment]
> Implementing a method from another interface doesn't seem to be disallowed,

It's not disallowed by the compiler, and it might make sense in application code (or with non-public interfaces); but avoiding the diamond is the only proto-guideline we have around DIMs for libraries.

**Comment by kimsey0** [comment]
OK. I that case, if `IHost` is to implement `IAsyncDisposable`, it will have to be without a default implementation.

**Comment by deeprobin** [comment]
Can we mark this as ready for review or are still things unclear?

**Comment by eerhardt** [comment]
It’s still a breaking change.

**Comment by maryamariyan** [comment]
We met again offline and decided to close this because the value here does not out-weight the cost of the breaking change being introduced.

**Comment by davidfowl** [comment]
@jaredpar this feels like a duck typing feature would be useful here where we could add an extension method for DisposeAsync and it would be enough for this type to support using statements (same for disposable)

**Comment by kimsey0** [comment]
@davidfowl, do you mean something like this?
```csharp
public static class HostExtensions
{
    public static ValueTask DisposeAsync(this IHost host)
    {
        if (host is IAsyncDisposable asyncDisposable)
        {
            return asyncDisposable.DisposeAsync();
        }
        else
        {
            host.Dispose();
            return ValueTask.CompletedTask;
        }
    }
}
```
